c-bindings-gen/src/types.rs

```diff
@@ -0,0 +1,1102 @@
+use std::collections::HashMap;
+
+use proc_macro2::{TokenTree, Span};
+
+// We have a few main types -
+// * Unit Enums, which are "mirrored" by creating a #[repr(C)] version and exposing it.
+// * Traits, which map to a void* and a jump table,
+// * Opaque Structs, which are exposed as-is either directly or via a typedef if they are generic
+//   (using our own concrete traits, which are enumerated below).
+//   All pub methods are exposed and pub fields are exposed via geters/setters.
```

Comment:
It would be helpful for reviewing this and maintenance if you wrote out minimal examples showing these transformations. Kudos if they can be doc tests. :)

Also, should this be placed in main.rs?